test/framework/control_plane.go

```diff
@@ -0,0 +1,225 @@
+		}
+		return cluster.Status.Phase
+	}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.ClusterPhaseProvisioned)))
+
+	// wait for all the machines to be running
+	By("waiting for all machines to be running")
+	for _, node := range input.Nodes {
+		Eventually(func() string {
+			machine := &clusterv1.Machine{}
+			key := client.ObjectKey{
+				Namespace: node.Machine.GetNamespace(),
+				Name:      node.Machine.GetName(),
+			}
+			if err := mgmtClient.Get(ctx, key, machine); err != nil {
+				return err.Error()
+			}
+			return machine.Status.Phase
+		}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.MachinePhaseRunning)))
+	}
+
+	By("waiting for the workload nodes to exist")
+	Eventually(func() []v1.Node {
+		nodes := v1.NodeList{}
+		By("ensuring the workload client can be generated")
+		err := wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
+			_, err := input.Management.GetWorkloadClient(ctx, input.Cluster.Namespace, input.Cluster.Name)
+			switch {
+			case apierrors.IsNotFound(err):
+				return false, nil
+			case err != nil:
+				return true, err
+			default:
+				return true, nil
+			}
+		})
+		Expect(err).NotTo(HaveOccurred())
+		By("getting the workload client and listing the nodes")
+		workloadClient, err := input.Management.GetWorkloadClient(ctx, input.Cluster.Namespace, input.Cluster.Name)
+		Expect(err).NotTo(HaveOccurred(), "Stack:\n%+v\n", err)
+		Expect(workloadClient.List(ctx, &nodes)).To(Succeed())
+		return nodes.Items
+	}, input.CreateTimeout, 10*time.Second).Should(HaveLen(len(input.Nodes)))
+}
+
+// CleanUp deletes the cluster and waits for everything to be gone.
+// Assertions:
+//   * Deletes Machines
+//   * Deletes MachineSets
+//   * Deletes MachineDeployments
+//   * Deletes KubeadmConfigs
+//   * Deletes InfraCluster
+//   * Deletes InfraMachines
+//   * Deletes InfraMachineTemplates
+//   * Deletes Secrets
+func (input *ControlplaneClusterInput) CleanUp() {
+	input.SetDefaults()
+	ctx := context.Background()
+	mgmtClient, err := input.Management.GetClient()
+	Expect(err).NotTo(HaveOccurred(), "stack: %+v", err)
+
+	By(fmt.Sprintf("deleting cluster %s", input.Cluster.GetName()))
+	Expect(mgmtClient.Delete(ctx, input.Cluster)).To(Succeed())
+
+	Eventually(func() []clusterv1.Cluster {
+		clusters := clusterv1.ClusterList{}
+		Expect(mgmtClient.List(ctx, &clusters)).To(Succeed())
+		return clusters.Items
+	}, input.DeleteTimeout, 10*time.Second).Should(HaveLen(0))
+
+	lbl, err := labels.Parse(fmt.Sprintf("%s=%s", clusterv1.ClusterLabelName, input.Cluster.GetClusterName()))
+	Expect(err).ToNot(HaveOccurred())
+	listOpts := &client.ListOptions{LabelSelector: lbl}
+
+	By("ensuring all artifacts have been deleted")
+	ensureArtifactsDeleted(ctx, mgmtClient, listOpts)
+
+	// TODO: Add more provider specific checks here.
+	switch TypeToKind(input.InfraCluster) {
+	case "DockerCluster":
+		ensureDockerArtifactsDeleted(ctx, mgmtClient, listOpts)
+	}
```

Comment:
I'll be adding support for `AWSCluster` as part of this story https://github.com/kubernetes-sigs/cluster-api/issues/1775